### PR TITLE
Enhance interpretation of GLOO_SOCKET_IFNAME

### PR DESCRIPTION
**Pull request by pietern**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#22978 Enhance interpretation of GLOO_SOCKET_IFNAME**

With this change you can now list multiple interfaces separated by
comma. ProcessGroupGloo creates a single Gloo context for every device
in the list (a context represents a connection to every other
rank). For every collective that is called, it will select the context
in a round robin fashion. The number of worker threads responsible for
executing the collectives is set to be twice the number of devices.

If you have a single physical interface, and wish to employ increased
parallelism, you can also specify
`GLOO_SOCKET_IFNAME=eth0,eth0,eth0,eth0`.  This makes ProcessGroupGloo
use 4 connections per rank, 4 I/O threads, and 8 worker threads
responsible for executing the collectives.

Differential Revision: [D16339962](https://our.internmc.facebook.com/intern/diff/D16339962/)

### CODE REVIEW

**Comment by ailzhang** [comment]
FYI here are the benchmark numbers I got from GCP, when scaling up number of connections between 2 hosts.

It does the trick that achieves full 16Gbps by adding more connections in Gloo :D But not sure if the non-linear scaling is concerning or not. 

But this PR is definitely helpful for users to optimize for their own network. Thanks @pietern ! 


|# connections | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8
-- | -- | -- | -- | -- | -- | -- | -- | --
Time to all_reduce 1 resnet50 | 0.216s | 0.148s | 0.172s | 0.0826s | 0.092s | 0.098s | 0.0702s | 0.075s

**Review by mrshenli — COMMENTED** [review]
Thanks for adding this optimization! Is it possible to add some tests for it?

File: torch/csrc/distributed/c10d/init.cpp [hunk 1]
```diff
@@ -427,26 +459,19 @@ They are used in specifying strategies for reduction collectives, e.g.,
                       int size,
                       std::chrono::milliseconds timeout) {
             ::c10d::ProcessGroupGloo::Options options;
-            ::gloo::transport::tcp::attr attr;
-            // First step, check "GLOO_SOCKET_IFNAME" environmental variable
-            // that can be set by the user
+
+            // Use interfaces listed in "GLOO_SOCKET_IFNAME", if set.
```
[the remark this answers is not in the document]
Comment (pietern) [reply torch/csrc/distributed/c10d/init.cpp]: Yes, good point.

File: torch/lib/c10d/ProcessGroupGloo.cpp [hunk 2]
```diff
@@ -329,6 +329,10 @@ uint32_t ProcessGroupGloo::nextTag() {
   return collectiveCounter_++;
 }
 
+std::shared_ptr<::gloo::Context> ProcessGroupGloo::getContext(uint32_t tag) {
+  return contexts_[tag % contexts_.size()];
```
[the remark this answers is not in the document]
Comment (pietern) [reply torch/lib/c10d/ProcessGroupGloo.cpp]: I added a note to the `torch.distributed` docs that dispatch is round robin.

File: torch/csrc/distributed/c10d/init.cpp [hunk 3]
```diff
@@ -35,8 +35,40 @@ namespace {
...
+std::shared_ptr<::gloo::transport::Device> createDeviceForDefaultHostname() {
+  ::gloo::transport::tcp::attr attr;
+
+  // Use the hostname to resolve the network address to
+  // use. Note: if the hostname does not resolve to an address (e.g.
+  // because of misconfigured /etc/hosts file), this will not work.
+  std::array<char, HOST_NAME_MAX> hostname{};
+  auto rv = gethostname(hostname.data(), hostname.size());
+  if (rv != 0) {
+    throw std::system_error(errno, std::system_category());
```
[the remark this answers is not in the document]
Comment (pietern) [reply torch/csrc/distributed/c10d/init.cpp]: I copied the routine from elsewhere in the file and this hasn't shown up as a problem before. I think we can leave it like this until it does.

**Review by mrshenli — APPROVED** [review]
LGTM!

**Pull request closed without merging**